### PR TITLE
DR-897 graceful shutdown

### PR DESCRIPTION
**Pull request by ddietterich**
Changes for quiet shutdown
- implement Unauthenticated /shutdown target (required helm & skaffold changes)
- teach gradle to use a local stairway jar for faster development
- add properties for shutdown timeout and connection pool controls
- lots of InterruptException throw declarations
- test to catch flights in all states

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -130,8 +130,19 @@ dependencies {
 
     compile 'org.springframework:spring-jdbc:5.1.9.RELEASE'
     compile 'org.broadinstitute.dsde.workbench:sam-client_2.12:0.1-9435410-SNAP'
-    compile 'bio.terra:stairway:0.0.3-SNAPSHOT'
     compile 'org.antlr:ST4:4.3'                          // String templating
+    compile group: 'io.kubernetes', name: 'client-java', version: '8.0.0'
+
+    // This allows you to build with a local Stairway JAR file, instead of the one from the repository
```
Comment (fboulnois) [inline build.gradle:136]: 👍

File: src/main/java/bio/terra/app/configuration/JdbcConfiguration.java [hunk 2]
```diff
@@ -20,6 +20,17 @@
     private String username;
     private String password;
     private String changesetFile;
+    /**
+     * Maximum number of database connections in the connection pool; -1 means no limit
```
Comment (fboulnois) [inline src/main/java/bio/terra/app/configuration/JdbcConfiguration.java:24]: Are there any thoughts about how the number of database connections will impact performance? I know Postgres in particular has issues with scaling as the number of database connections increases (say >100 total connections)

File: skaffold.yaml.template [hunk 3]
```diff
@@ -25,8 +25,8 @@ deploy:
       valuesFiles:
       - https://raw.githubusercontent.com/broadinstitute/datarepo-helm-definitions/master/dev/TEMP/TEMPSecrets.yaml
     - name: TEMP-jade
-      chartPath: https://github.com/broadinstitute/datarepo-helm/releases/download/datarepo-0.1.0/datarepo-0.1.0.tgz
-      version: 0.1.0
+      chartPath: https://github.com/broadinstitute/datarepo-helm/releases/download/datarepo-0.1.1/datarepo-0.1.1.tgz
+      version: 0.1.1
```
Comment (fboulnois) [inline skaffold.yaml.template:29]: Can you comment in more detail on how the the skaffold changes enable shutdown?

File: src/main/java/bio/terra/service/job/JobService.java [hunk 4]
```diff
@@ -60,10 +64,28 @@ public JobService(IamService samService,
...
+    /**
+     * Stop accepting jobs and shutdown stairway
+     */
+    public void shutdown() throws InterruptedException {
+        isShutdown.set(true);
+
+        // We allocate 3/4 of the time for graceful shutdown. Then call terminate for the rest of the time.
+        int shutdownTimeout = appConfig.getShutdownTimeoutSeconds();
+        int gracefulTimeout = (shutdownTimeout * 3) / 4;
+        int terminateTimeout = (shutdownTimeout - gracefulTimeout) - 2;
```
Comment (fboulnois) [inline src/main/java/bio/terra/service/job/JobService.java:95]: What happens if `terminateTimeout` < 0?

**Review by fboulnois — CHANGES_REQUESTED** [review]
A few comments / questions:

File: src/main/java/bio/terra/service/job/JobService.java [hunk 5]
```diff
@@ -60,10 +64,28 @@ public JobService(IamService samService,
...
         ExceptionSerializer serializer = new StairwayExceptionSerializer(objectMapper);
-        stairway = new Stairway(executorService, applicationContext, serializer);
+        stairway = new Stairway(executorService, applicationContext, serializer, "stairwayname");
+    }
+
+    /**
+     * Stop accepting jobs and shutdown stairway
+     */
+    public void shutdown() throws InterruptedException {
+        isShutdown.set(true);
```
Comment (marikomedlock) [inline src/main/java/bio/terra/service/job/JobService.java:84]: Do you need a check of the atomic boolean at the beginning of this method to make sure that only one thread is calling shutdown() at a time?

**Review by marikomedlock — APPROVED** [review]
LGTM. One question about multiple threads calling shutdown at the same time. Thanks for your patience and apologies for the delay.

File: src/main/java/bio/terra/app/configuration/JdbcConfiguration.java (hunk 2 above)
Comment (ddietterich) [reply src/main/java/bio/terra/app/configuration/JdbcConfiguration.java:24]: I'm not familiar with an inherent performance limitation. Once active work has used all of the system resources, adding work will slow things down.

In general, we do not run long transactions, so connection pooling should work well for us.

File: skaffold.yaml.template (hunk 3 above)
Comment (ddietterich) [reply skaffold.yaml.template:29]: We have a screwy path for unauthenticated REST API paths. They are routed through the UI's nginx and from there are routed to the DRmanager endpoint.

In order to add a new unauthenticated REST API, I had to change the UI deployment. And in order to pick up that deployment change, I had to bump the version number in several places. This one makes sure we use the right helm stuff when deploying via skaffold.

File: src/main/java/bio/terra/service/job/JobService.java (hunk 5 above)
Comment (ddietterich) [reply src/main/java/bio/terra/service/job/JobService.java:84]: Good point. Yes, we should ignore duplicate shutdown requests.

File: src/main/java/bio/terra/service/job/JobService.java (hunk 4 above)
Comment (ddietterich) [reply src/main/java/bio/terra/service/job/JobService.java:95]: Good point. I added a floor, so we will never use too small a value.

File: src/main/java/bio/terra/app/configuration/JdbcConfiguration.java (hunk 2 above)
Comment (fboulnois) [reply src/main/java/bio/terra/app/configuration/JdbcConfiguration.java:24]: We've also discussed how we might address this in the scaling notes, so I'll close this out.

File: src/main/java/bio/terra/service/job/JobService.java [hunk 6]
```diff
@@ -75,10 +77,19 @@ public JobService(IamService samService,
...
+            logger.warn("Ignoreing duplicate shutdown request");
+            return;
+        }
         isShutdown.set(true);
 
+        // We enforce a minimum shutdown time. Otherwise, there is no point in trying the shutdown.
         // We allocate 3/4 of the time for graceful shutdown. Then call terminate for the rest of the time.
         int shutdownTimeout = appConfig.getShutdownTimeoutSeconds();
+        if (shutdownTimeout < MIN_SHUTDOWN_TIMEOUT) {
+            logger.warn("Shutdown timeout of " + shutdownTimeout + "is too small. Setting to " + MIN_SHUTDOWN_TIMEOUT);
```
Comment (fboulnois) [inline src/main/java/bio/terra/service/job/JobService.java:90]:
```suggestion
            logger.warn("Shutdown timeout of " + shutdownTimeout + "is too small. Setting to " + MIN_SHUTDOWN_TIMEOUT);
            shutdownTimeout = MIN_SHUTDOWN_TIMEOUT;
```

**Review by fboulnois — CHANGES_REQUESTED** [review]
Edit: see below.

File: src/main/java/bio/terra/service/job/JobService.java [hunk 7]
```diff
@@ -60,10 +66,37 @@ public JobService(IamService samService,
...
+            return;
+        }
+        isShutdown.set(true);
+
+        // We enforce a minimum shutdown time. Otherwise, there is no point in trying the shutdown.
+        // We allocate 3/4 of the time for graceful shutdown. Then call terminate for the rest of the time.
+        int shutdownTimeout = appConfig.getShutdownTimeoutSeconds();
+        if (shutdownTimeout < MIN_SHUTDOWN_TIMEOUT) {
+            logger.warn("Shutdown timeout of " + shutdownTimeout + "is too small. Setting to " + MIN_SHUTDOWN_TIMEOUT);
+        }
```
Comment (fboulnois) [inline src/main/java/bio/terra/service/job/JobService.java]:
```suggestion
        int shutdownTimeout = appConfig.getShutdownTimeoutSeconds();
        if (shutdownTimeout < MIN_SHUTDOWN_TIMEOUT) {
            logger.warn("Shutdown timeout of " + shutdownTimeout + "is too small. Setting to " + MIN_SHUTDOWN_TIMEOUT);
            shutdownTimeout = MIN_SHUTDOWN_TIMEOUT;
        }
```

**Review by fboulnois — CHANGES_REQUESTED** [review]
I think the floor is missing logic, otherwise lgtm:

File: src/main/java/bio/terra/service/job/JobService.java (hunk 7 above)
Comment (ddietterich) [reply src/main/java/bio/terra/service/job/JobService.java]: Oh duh! Thank you

**Review by fboulnois — APPROVED** [review]
lgtm now 👍

**Pull request merged**